File: app/src/main/java/com/huanli233/biliterminal2/activity/article/ArticleInfoActivity.java
package com.huanli233.biliterminal2.activity.article;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;

import com.huanli233.biliterminal2.R;
import com.huanli233.biliterminal2.activity.base.BaseActivity;
import com.huanli233.biliterminal2.activity.reply.ReplyFragment;
import com.huanli233.biliterminal2.adapter.viewpager.ViewPagerFragmentAdapter;
import com.huanli233.biliterminal2.api.ReplyApi;
import com.huanli233.biliterminal2.event.ReplyEvent;
import com.huanli233.biliterminal2.helper.TutorialHelper;
import com.huanli233.biliterminal2.util.AnimationUtils;
import com.huanli233.biliterminal2.util.MsgUtil;
import com.huanli233.biliterminal2.util.TerminalContext;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.ArrayList;
import java.util.List;

public class ArticleInfoActivity extends BaseActivity {
    private long cvid;

    private ReplyFragment replyFragment;
    private long seek_reply;

    private ImageView loadingView;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_simple_viewpager);
        Intent intent = getIntent();
        cvid = intent.getLongExtra("cvid", 114514);
        this.seek_reply = getIntent().getLongExtra("seekReply", -1);

        setPageName("专栏详情");
        loadingView = findViewById(R.id.loading);

        TutorialHelper.showTutorialList(this, R.array.tutorial_article, 7);

        ViewPager viewPager = findViewById(R.id.viewPager);

        TerminalContext.getInstance().getArticleInfoByCvId(cvid)
            .observe(this, (result) -> result.onSuccess((articleInfo)-> {
                List<Fragment> fragmentList = new ArrayList<>();
                ArticleInfoFragment articleInfoFragment = ArticleInfoFragment.newInstance(cvid);
                fragmentList.add(articleInfoFragment);
                replyFragment = ReplyFragment.newInstance(cvid, ReplyApi.REPLY_TYPE_ARTICLE, seek_reply, articleInfo != null ? articleInfo.upInfo.mid : -1);
                replyFragment.setSource(articleInfo);
                fragmentList.add(replyFragment);
                ViewPagerFragmentAdapter vpfAdapter = new ViewPagerFragmentAdapter(getSupportFragmentManager(), fragmentList);
                viewPager.setAdapter(vpfAdapter);
                View view;
                if ((view = articleInfoFragment.getView()) != null)
                    view.setVisibility(View.GONE);
                if (seek_reply != -1) viewPager.setCurrentItem(1);
                articleInfoFragment.setOnFinishLoad(() -> AnimationUtils.crossFade(loadingView, articleInfoFragment.getView()));
                TutorialHelper.showPagerTutorial(this,2);
            }).onFailure((error) -> {
                loadingView.setImageResource(R.mipmap.loading_2233_error);
                MsgUtil.err(error);
            }));
    }

    @Override
    protected boolean eventBusEnabled() {
        return true;
    }

    @Subscribe(threadMode = ThreadMode.ASYNC, sticky = true, priority = 1)
    public void onEvent(ReplyEvent event) {
        replyFragment.notifyReplyInserted(event);
    }

    @Override
    protected void onDestroy() {
        TerminalContext.getInstance().leaveDetailPage();
        super.onDestroy();
    }
}

File: app/src/main/java/com/huanli233/biliterminal2/activity/article/ArticleInfoActivity.kt
package com.huanli233.biliterminal2.activity.article

import android.annotation.SuppressLint
import android.content.Intent
import android.os.Bundle
import androidx.activity.viewModels
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.huanli233.biliterminal2.R
import com.huanli233.biliterminal2.activity.base.BaseActivity
import com.huanli233.biliterminal2.activity.opus.OpusActivity
import com.huanli233.biliterminal2.activity.opus.OpusViewModel
import com.huanli233.biliterminal2.api.apiResultNonNull
import com.huanli233.biliterminal2.api.bilibiliApi
import com.huanli233.biliterminal2.api.toResultNonNull
import com.huanli233.biliterminal2.databinding.ActivitySimpleViewpagerBinding
import com.huanli233.biliterminal2.util.extensions.LoadState
import com.huanli233.biliterminal2.util.extensions.invisible
import com.huanli233.biliterminal2.util.extensions.showError
import com.huanli233.biliterminal2.util.extensions.visible
import com.huanli233.biliwebapi.api.interfaces.IArticleApi
import com.huanli233.biliwebapi.bean.article.ArticleInfo
import com.huanli233.biliwebapi.bean.opus.Opus
import kotlinx.coroutines.launch
import kotlin.getValue

class ArticleInfoActivity : BaseActivity() {

    val viewModel: ArticleInfoViewModel by viewModels {
        ArticleInfoViewModelFactory(intent.getLongExtra("cvid", -1))
    }

    @SuppressLint("MissingInflatedId")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding = ActivitySimpleViewpagerBinding.inflate(layoutInflater)

        setPageName(getString(R.string.article_detail))
        binding.loading.visible()
        binding.viewPager.invisible()

        viewModel.data.observe(this) {
            when (it) {
                is LoadState.Success -> {
                    startActivity(
                        Intent(this, OpusActivity::class.java).apply {
                            putExtra("opusId", it.data.dynIdStr)
                            intent.extras?.let { extras -> putExtras(extras) }
                        }
                    )
                    finish()
                }

                is LoadState.Error -> {
                    binding.loading.showError()
                }

                else -> Unit
            }
        }
    }
}

@Suppress("DEPRECATION")
class ArticleInfoViewModel(val cvid: Long): ViewModel() {

    private val _data = MutableLiveData<LoadState<ArticleInfo>>()
    val data: LiveData<LoadState<ArticleInfo>> get() = _data

    init {
        fetchData()
    }

    fun fetchData() {
        viewModelScope.launch {
            bilibiliApi.api(IArticleApi::class) { getArticle(cvid) }
                .apiResultNonNull()
                .onSuccess {
                    _data.postValue(LoadState.Success(it))
                }.onFailure {
                    _data.postValue(LoadState.Error(it))
                }
        }
    }

}

class ArticleInfoViewModelFactory(private val cvid: Long) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(OpusViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return ArticleInfoViewModel(cvid) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}